File: Etapa3-Construcao/3.1-Stack1_Node_TS/src/domain/entities/Hospede.ts
import { CPF } from '../value-objects/CPF.js';
import { Email } from '../value-objects/Email.js';

// Entidade Hóspede com VOs para CPF e Email, garantindo consistência.
export class Hospede {
  private _id?: string;
  private _nome: string;
  private _sobrenome: string;
  private _cpf: CPF;
  private _email: Email;
  private _ativo = true;

  private constructor(params: { id?: string; nome: string; sobrenome: string; cpf: CPF; email: Email }) {
    this._id = params.id;
    this._nome = Hospede.validarNome(params.nome);
    this._sobrenome = Hospede.validarNome(params.sobrenome);
    this._cpf = params.cpf;
    this._email = params.email;
  }

  static criar(params: { id?: string; nome: string; sobrenome: string; cpf: CPF; email: Email }): Hospede {
    return new Hospede(params);
  }

  atualizar(parciais: Partial<{ nome: string; sobrenome: string; email: Email }>): void {
    if (parciais.nome !== undefined) this._nome = Hospede.validarNome(parciais.nome);
    if (parciais.sobrenome !== undefined) this._sobrenome = Hospede.validarNome(parciais.sobrenome);
    if (parciais.email !== undefined) this._email = parciais.email;
  }

  desativar(): void { this._ativo = false; }

  private static validarNome(v: string): string {
    const t = v.trim();
    if (t.length < 2) throw new Error('Nome inválido');
    return t;
  }

  get id(): string | undefined { return this._id; }
  get nome(): string { return this._nome; }
  get sobrenome(): string { return this._sobrenome; }
  get cpf(): CPF { return this._cpf; }
  get email(): Email { return this._email; }
  get ativo(): boolean { return this._ativo; }
}

File: Etapa3-Construcao/3.2-Stack2_Java_Spring/src/main/java/com/hotel/hospedes/domain/Hospede.java
package com.hotel.hospedes.domain;

import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "hospede", uniqueConstraints = @UniqueConstraint(columnNames = "cpf"))
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Hospede {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false)
    private String sobrenome;

    @Column(nullable = false, length = 11)
    private String cpf; // Validação de DV pode ser aplicada via @Pattern ou serviço

    @Column(nullable = false)
    private String email; // Validação @Email em DTO

    @Column(nullable = false)
    private boolean ativo = true;

    @Builder
    private Hospede(String nome, String sobrenome, String cpf, String email) {
        this.nome = validarNome(nome);
        this.sobrenome = validarNome(sobrenome);
        this.cpf = cpf; // Simplificado; regra mais rígida pode ser aplicada em VO/validador
        this.email = email;
    }

    public void atualizar(String nome, String sobrenome, String email) {
        if (nome != null) this.nome = validarNome(nome);
        if (sobrenome != null) this.sobrenome = validarNome(sobrenome);
        if (email != null) this.email = email;
    }

    public void desativar() { this.ativo = false; }

    private static String validarNome(String v) {
        if (v == null || v.trim().length() < 2) throw new IllegalArgumentException("Nome inválido");
        return v.trim();
    }
}
